### DISCUSSION TITLE
Convert audio file to image file and back using raw headerless binary blobs

### DISCUSSION THREAD

**Discussion by iwanPlays** [ImageMagick/ImageMagick#2076] [Help] [answered]
Hello, I would like to have some fun by editing an audio file in an image editor.

Based on https://stackoverflow.com/questions/2059014/converting-raw-audio-data-to-wav-with-scripting I'd convert the audio file to raw audio data (.raw being a headerless format in the audio world) using sox:

    sox -r 44100 -b 16 -c 2 sound.wav -e unsigned -b 16 -c 2 raw.raw

I then try to use imagemagick to interpret it as an rgba file, which according to http://www.imagemagick.org/discourse-server/viewtopic.php?t=22755 is headerless:

    mv raw.raw raw.rgba
    magick convert -size 500x500 -depth 32 raw.rgba test.png

I picked 500x500 at random, being unsure how to calculate how large a square the data could actually fill and being unsure how to make imagemagick fill missing data with zeroes. Unfortunately, imagemagick 7.09-26 Q16 x64 win10 doesn't like this:

    convert: unexpected end-of-file 'raw.rgba': No such file or directory @ error/rgb.c/ReadRGBImage/247.
    convert: no images defined `test.png' @ error/convert.c/ConvertImageCommand/3282.

Any ideas what my mistake might be or how to reach the goal?

**Comment by fmw42** [comment]
I am not an expert on doing what you want, but ImageMagick will object if the -size dimensions and the -depth value are not precise for your data. I do not think there is a built-in method to tell it to fill to that size. But why rgba rather than just rgb or probably better as single channel gray?

**Comment by snibgo** [comment]
You can create a Nx1 or Nx2 pixel file from raw sound. See [Sound and vision](http://im.snibgo.com/sndvis.htm]). The IM depth is bits per channel per pixel, not total bits per channel.

**Comment by iwanPlays** [comment, marked as answer]
Thank you both!

> But why rgba rather than just rgb or probably better as single channel gray?

I thought it might look cool if there's transparency. I switched to RGB in my attempt because I'm a little more used to thinking about RBG binary. I'll try .gray next.

> The IM depth is bits per channel per pixel, not total bits per channel.

This actually got me thinking enough to figure out a working solution!

```
# wav2rawpng.sh
infile=$1
sox -r 44100 -b 16 -c 2 $infile -e unsigned -b 16 -c 2 test.raw
file=test.rgb
mv test.raw $file
# find out size in bytes
size=$(du -b $file)
echo size in bytes is $size
# calculate padding needed to ensure square image
# I don't even know what's going on with the 6's...
pad=$(awk -v size="$size" 'function quad(x){return(x*x)} function ceil(x, y){y=int(x); return(x>y?y+1:y)} BEGIN{ print quad(6*ceil(sqrt(size/6)))/6-size }')
echo padding in bytes is $pad
cp test.rgb test-pad.rgb
truncate -s +$pad test-pad.rgb

# oh yeah we need the resolution too
newedge=$(awk -v size="$size" 'function ceil(x, y){y=int(x); return(x>y?y+1:y)} BEGIN{ print ceil(sqrt(size/6)) }')
echo edge size is $newedge
magick convert -size ${newedge}x${newedge} -depth 16 test-pad.rgb ${infile%.wav}-rawpng.png
```

... then edit $infile-rawpng.png in GIMP for example and ...

```
# rawpng2wav.sh
infile=$1
magick convert $infile -depth 16 test2.rgb
mv test2.rgb test2.raw
sox -e unsigned -b 16 -c 2 -r 44100 test2.raw -r 44100 -b 16 -c 2 ${infile%.png}.wav
```

I run these in Git Bash on Win10. Easiest way to install bash on Windows in my humble opinion.

Notes:
- Distorts (Lens Distortion) is fun.
- Color balance adds a horrible base whine.
- I'm hoping that .gray might be even more fun to edit.

Here's an example of an output: https://freesound.org/people/iwanPlays/sounds/520483/

I'm guessing right now I'm converting multiple samples into one pixel, making any edit likely to distort. Using .gray will solve this, right? I'll have to sleep over this but I warmly welcome suggestions :)

**Comment by duangsuse** [comment]
I tried to port wav2rawpng.sh to numpy, it works, but converting .rgb(48000 Hz, 2 channels, s16,) back to wav breaks the audio.

N,pad,l=(1478996, 3058, 497)
rawpng.png
![a-rawpng](https://github.com/ImageMagick/ImageMagick/assets/10570123/dfecfd13-95a1-4c44-88dc-c217c883a01b)


```py
import numpy as np
import subprocess
from PIL import Image

sh=lambda s:subprocess.run(s, shell=True)
infile = 'a.wav'; sr=48000
outfile=f'{infile[:-4]}-rawpng.png'

# Convert WAV to raw data using sox
sh(f'sox -r {sr} -b 16 -c 2 {infile} -e unsigned -b 16 -c 2 test.raw')

# Read the raw data into a NumPy array
raw = np.fromfile('test.raw', dtype=np.uint8)
N = raw.nbytes

# Calculate padding
l = int(np.ceil(np.sqrt(N / 6)))
pad = (l ** 2) * 6 - N
print(f"{N,pad,l=}")

# Apply padding to the raw data and save it to a new fiimport numpy as np
import subprocess
from PIL import Image

sh=lambda s:subprocess.run(s, shell=True)
infile = 'a.wav'; sr=48000
outfile=f'{infile[:-4]}-rawpng.png'

# Convert WAV to raw data using sox
sh(f'sox -r {sr} -b 16 -c 2 {infile} -e unsigned -b 16 -c 2 test.raw')

# Read the raw data into a NumPy array
raw = np.fromfile('test.raw', dtype=np.uint8)
N = raw.nbytes

# Calculate padding
l = int(np.ceil(np.sqrt(N / 6)))
pad = (l ** 2) * 6 - N
print(f"{N,pad,l=}")

# Apply padding to the raw data and save it to a new file
img = np.pad(raw, (0, pad), 'constant')
img.tofile('test.rgb')

# Convert the padded data to a RAW PNG image
sh(f'convert -size {l}x{l} -depth 16 test.rgb {outfile}')

# Create an image from raw RGB data
img = np.pad(raw, (0, pad), 'constant')
img.tofile('test.rgb')

# Convert the padded data to a RAW PNG image
sh(f'convert -size {l}x{l} -depth 16 test.rgb {outfile}')

# Create an image from raw RGB data
Image.frombytes('I;16', (l, l), img)

##
import IPython.display as say
from scipy.io import wavfile

# ffplay -f s16le -ac 2 -ar 48000 test.rgb
wavfile.write('a1.wav', sr, img[:img.size - pad].astype(np.int16).reshape(-1, 2))
say.Audio('a1.wav')
```

**Comment by duangsuse** [comment]
btw. it's better to make a wav2png using `librosa.stft` [see also FFT](https://pythonnumericalmethods.berkeley.edu/notebooks/chapter24.02-Discrete-Fourier-Transform.html)
![badapple.wav](https://github.com/ImageMagick/ImageMagick/assets/10570123/48317750-c9e8-489d-90a1-be3283578eae)


```py
import \
  librosa as sa, soundfile as sf ,\
  numpy as np, IPython.display as ipd

from PIL import Image
from PIL import ImageFilter as Fi, ImageGrab as Grab

w2bmp=lambda k:pipe(wav(k), ft, (spA,spB), (np.flipud,np.flipud), bmp('HSV') )

pipe=lambda*eq: reduce(lambda a,b:\
  (lambda x:b[0](a[0](x)), lambda y:a[1](b[1](y))), eq)

def wav(f,sr=48000,ver=''):
  def to(v): nonlocal ver;ver=v; return librosa.load(f, sr=sr)[0]
  return to, lambda y: sf.write(ver+f,y, sr,'PCM_16')

bmp=lambda rgb: (
  lambda F:Image.fromarray(F.astype(np.uint8), mode=rgb).convert("RGB"),
  lambda im:np.array(im.convert(rgb)).astype(float)
)

def spA(s):
  V,S=s; H=S.copy();H.fill(0)
  return np.dstack((H, S, V)) # 0, power, phrase
def spB(a):
  H,S,V=(a[:,:,0] for a in np.dsplit(a,3))
  return V,S

yMax=300
ft=(lambda y: (F:=sa.stft(y)[:yMax], (np.abs(F),np.angle(F)) )[-1] ,\
    lambda px:(lambda V,S:istft(V * np.exp(1j * S) ))(*(np.pad(x, [(0, 1025-yMax)], 'constant') for x in px)) )

try:
    cat,cut=w2bmp('a.wav')
    a=cat('out_'); cut(a)
    b=cat('fi_').filter(Fi.GaussianBlur(radius=3)); cut(b)
    display(a, ipd.Audio('out_a.wav'), b, ipd.Audio('fi_a.wav'))
except:
    cut(Grab.grabclipboard())
    ipd.Audio('a.wav')
```

[src](https://gist.github.com/duangsuse/3aa22e85b4417a6e6618204b6ab8bfaf#file-note-ipynb)
